lib/core/src/Cardano/Pool/Metadata.hs

```diff
@@ -251,26 +263,37 @@ fetchStakePoolMetaZip
     -> IO (Either FetchError FilePath)
 fetchStakePoolMetaZip tr url zipFileName = do
     traceWith tr MsgDownloadStarted
-    fmap join $ tryJust fileExceptionHandler $
+    res <- flip onException cleanup $ fmap join $ tryJust fileExceptionHandler $
         withFile zipFileName WriteMode $ \zipFile -> tryJust handler $ do
             req <- parseUrlThrow url
             httpSink req $ \_response -> Conduit.mapM_ (BS.hPut zipFile)
             size <- hTell zipFile
             traceWith tr $ MsgDownloadComplete size
             pure zipFileName
+    when (isLeft res) cleanup
+    pure res
```

Comment:
It seems redundant to catch exceptions at 3 different places here and, capture some of them at the value-level while others remains interruptions. 
What about using a more "uniform" approach and, get the "unknown exception" as part of the `FetchError` data-type?

Reply:
Yeah I know what you mean. Multiple handlers, apparent mixing of checked and unchecked exceptions, two cleanup paths. I tried to come up with something simpler but -

I want to differentiate between the exception types so that the error messages can describe what it was doing to cause the exception.

As far as I can tell, the only normal exceptions we should get are IOException and HttpException. If we add an "unknown exception" constructor for the Left FetchError, that exception handler should not swallow async exceptions.

So we would still need both `onException cleanup` and `when (isLeft res) cleanup`. The first one for the async exceptions that are not swallowed.